libs/wire-api/src/Wire/API/MLS/Extension.hs

```diff
@@ -0,0 +1,144 @@
+-- ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS
+-- FOR A PARTICULAR PURPOSE. See the GNU Affero General Public License for more
+-- details.
+--
+-- You should have received a copy of the GNU Affero General Public License along
+-- with this program. If not, see <https://www.gnu.org/licenses/>.
+
+module Wire.API.MLS.Extension
+  ( -- * Extensions
+    Extension (..),
+    decodeExtension,
+    parseExtension,
+    ExtensionTag (..),
+    CapabilitiesExtensionTagSym0,
+    LifetimeExtensionTagSym0,
+    SExtensionTag (..),
+    SomeExtension (..),
+    Capabilities (..),
+    Lifetime (..),
+
+    -- * Other types
+    Timestamp (..),
+    ProtocolVersion (..),
+    ProtocolVersionTag (..),
+
+    -- * Utilities
+    pvTag,
+    tsPOSIX,
+  )
+where
+
+import Data.Binary
+import Data.Singletons.TH
+import Data.Time.Clock.POSIX
+import Imports
+import Wire.API.Arbitrary
+import Wire.API.MLS.CipherSuite
+import Wire.API.MLS.Serialisation
+
+newtype ProtocolVersion = ProtocolVersion {pvNumber :: Word8}
+  deriving newtype (Eq, Ord, Show, Binary, Arbitrary)
+  deriving (ParseMLS) via (BinaryMLS ProtocolVersion)
+
+data ProtocolVersionTag = ProtocolMLS10 | ProtocolMLSDraft11
+  deriving stock (Bounded, Enum, Eq, Show, Generic)
+  deriving (Arbitrary) via GenericUniform ProtocolVersionTag
+
+pvTag :: ProtocolVersion -> Maybe ProtocolVersionTag
+pvTag (ProtocolVersion v) = case v of
+  1 -> pure ProtocolMLS10
+  200 -> pure ProtocolMLSDraft11
+  _ -> Nothing
+
+data Extension = Extension
+  { extType :: Word16,
+    extData :: ByteString
+  }
+  deriving stock (Eq, Show, Generic)
+  deriving (Arbitrary) via GenericUniform Extension
+
+instance ParseMLS Extension where
+  parseMLS = Extension <$> parseMLS <*> parseMLSBytes @Word32
+
+data ExtensionTag
+  = CapabilitiesExtensionTag
+  | LifetimeExtensionTag
+  deriving (Bounded, Enum)
+
+$(genSingletons [''ExtensionTag])
+
+type family ExtensionType (t :: ExtensionTag) :: * where
+  ExtensionType 'CapabilitiesExtensionTag = Capabilities
+  ExtensionType 'LifetimeExtensionTag = Lifetime
+
+parseExtension :: Sing t -> Get (ExtensionType t)
+parseExtension SCapabilitiesExtensionTag = parseMLS
+parseExtension SLifetimeExtensionTag = parseMLS
+
+data SomeExtension where
+  SomeExtension :: Sing t -> ExtensionType t -> SomeExtension
+
+instance Eq SomeExtension where
+  SomeExtension SCapabilitiesExtensionTag caps1 == SomeExtension SCapabilitiesExtensionTag caps2 = caps1 == caps2
+  SomeExtension SLifetimeExtensionTag lt1 == SomeExtension SLifetimeExtensionTag lt2 = lt1 == lt2
+  _ == _ = False
+
+instance Show SomeExtension where
+  show (SomeExtension SCapabilitiesExtensionTag caps) = show caps
+  show (SomeExtension SLifetimeExtensionTag lt) = show lt
+
+decodeExtension :: Extension -> Either Text (Maybe SomeExtension)
+decodeExtension e = do
+  case toMLSEnum' (extType e) of
+    Left MLSEnumUnkonwn -> pure Nothing
```

Comment:
A typo in `Unkonwn`.

Reply:
Done in 5c301984a49c8f4227e77914fa45a3c4661c0c84.